File: Hb Track - Backend/app/api/v1/routers/positions.py
"""
Router para Posições (Defensivas e Ofensivas)
=============================================

Endpoints para listar posições disponíveis no sistema.
Usado nos formulários de cadastro de atletas.
"""

from typing import List

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from pydantic import BaseModel

from app.core.db import get_db
from app.models.defensive_position import DefensivePosition
from app.models.offensive_position import OffensivePosition


router = APIRouter(
    tags=["positions"],
)


class PositionResponse(BaseModel):
    """Schema de resposta para posição."""
    id: int
    name: str
    code: str
    
    class Config:
        from_attributes = True


class PositionsListResponse(BaseModel):
    """Schema de resposta com todas as posições."""
    defensive: List[PositionResponse]
    offensive: List[PositionResponse]


# =============================================================================
# GET /v1/positions - Listar todas as posições
# =============================================================================

@router.get(
    "",
    response_model=PositionsListResponse,
    summary="Listar posições",
    description="""
Retorna todas as posições defensivas e ofensivas disponíveis.

Usado nos formulários de cadastro de atletas.

**Regras:**
- RD13: Goleiras não podem ter posição ofensiva
- Posições defensivas: goleira, ponta, meia, pivô, central
- Posições ofensivas: ponta esquerda, ponta direita, armador central, armador esquerdo, armador direito, pivô
""",
)
async def list_positions(
    db: Session = Depends(get_db),
):
    """Lista todas as posições defensivas e ofensivas."""
    
    defensive = db.query(DefensivePosition).order_by(DefensivePosition.id).all()
    offensive = db.query(OffensivePosition).order_by(OffensivePosition.id).all()
    
    return PositionsListResponse(
        defensive=[PositionResponse.model_validate(p) for p in defensive],
        offensive=[PositionResponse.model_validate(p) for p in offensive],
    )


# =============================================================================
# GET /v1/positions/defensive - Listar posições defensivas
# =============================================================================

@router.get(
    "/defensive",
    response_model=List[PositionResponse],
    summary="Listar posições defensivas",
    description="Retorna todas as posições defensivas disponíveis.",
)
async def list_defensive_positions(
    db: Session = Depends(get_db),
):
    """Lista posições defensivas."""
    positions = db.query(DefensivePosition).order_by(DefensivePosition.id).all()
    return [PositionResponse.model_validate(p) for p in positions]


# =============================================================================
# GET /v1/positions/offensive - Listar posições ofensivas
# =============================================================================

@router.get(
    "/offensive",
    response_model=List[PositionResponse],
    summary="Listar posições ofensivas",
    description="Retorna todas as posições ofensivas disponíveis.",
)
async def list_offensive_positions(
    db: Session = Depends(get_db),
):
    """Lista posições ofensivas."""
    positions = db.query(OffensivePosition).order_by(OffensivePosition.id).all()
    return [PositionResponse.model_validate(p) for p in positions]

File: Hb Track - Frontend/lib/api/positions.ts
/**
 * Serviço de API - Positions (Posições Defensivas e Ofensivas)
 * 
 * Conforme REGRAS.md V1.2:
 * - RDB17: Tabelas de lookup são globais
 * - RD13: Goleiras (defensive_position_id=5) não podem ter posição ofensiva
 */

import { apiClient } from './client';

export interface DefensivePosition {
  id: number;
  code: string;
  name: string;
  abbreviation?: string;
  is_active: boolean;
}

export interface OffensivePosition {
  id: number;
  code: string;
  name: string;
  abbreviation?: string;
  is_active: boolean;
}

export interface SchoolingLevel {
  id: number;
  code: string;
  name: string;
  is_active: boolean;
}

export const defensivePositionsService = {
  /**
   * Lista todas as posições defensivas
   */
  async list(): Promise<DefensivePosition[]> {
    // Backend retorna array direto, não { items: [...] }
    return apiClient.get<DefensivePosition[]>('/defensive-positions');
  },

  async getById(id: number): Promise<DefensivePosition> {
    return apiClient.get<DefensivePosition>(`/defensive-positions/${id}`);
  },
};

export const offensivePositionsService = {
  /**
   * Lista todas as posições ofensivas
   */
  async list(): Promise<OffensivePosition[]> {
    // Backend retorna array direto, não { items: [...] }
    return apiClient.get<OffensivePosition[]>('/offensive-positions');
  },

  async getById(id: number): Promise<OffensivePosition> {
    return apiClient.get<OffensivePosition>(`/offensive-positions/${id}`);
  },
};

export const schoolingLevelsService = {
  /**
   * Lista todos os níveis de escolaridade
   */
  async list(): Promise<SchoolingLevel[]> {
    // Backend retorna array direto, não { items: [...] }
    return apiClient.get<SchoolingLevel[]>('/schooling-levels');
  },

  async getById(id: number): Promise<SchoolingLevel> {
    return apiClient.get<SchoolingLevel>(`/schooling-levels/${id}`);
  },
};
